drop-in/src/main/java/com/adyen/checkout/dropin/service/SessionDropInService.kt

```diff
@@ -0,0 +1,211 @@
+                            reason = it.message,
+                            dismissDropIn = false,
+                        )
+                        sendSessionSetupResult(result)
+                    }
+                )
+        }
+    }
+
+    private fun sendSessionSetupResult(sessionSetupDropInServiceResult: SessionSetupDropInServiceResult) {
+        Logger.d(TAG, "Sending session setup result")
+        emitResult(sessionSetupDropInServiceResult)
+    }
+
+    override fun onPaymentsCallRequested(
+        paymentComponentState: PaymentComponentState<*>,
+        paymentComponentJson: JSONObject
+    ) {
+        launch {
+            sessionRepository.submitPayment(paymentComponentState.data)
+                .fold(
+                    onSuccess = { response ->
+                        val action = response.action
+                        val result = when {
+                            action != null -> DropInServiceResult.Action(action)
+                            response.order.isNonFullyPaid() -> {
+                                updatePaymentMethods(response.order)
+                                null
+                            }
+                            else -> DropInServiceResult.Finished(response.resultCode ?: "EMPTY")
+                        }
+                        result?.let { sendResult(result) }
+                    },
+                    onFailure = ::onFailure
+                )
+        }
+    }
+
+    private fun OrderResponse?.isNonFullyPaid() = (this?.remainingAmount?.value ?: 0) > 0
+
+    override fun onDetailsCallRequested(actionComponentData: ActionComponentData, actionComponentJson: JSONObject) {
+        launch {
+            sessionRepository.submitDetails(actionComponentData)
+                .fold(
+                    onSuccess = { response ->
+                        val result = when (val action = response.action) {
+                            null -> DropInServiceResult.Finished(response.resultCode ?: "EMPTY")
+                            else -> DropInServiceResult.Action(action)
+                        }
+                        sendResult(result)
+                    },
+                    onFailure = ::onFailure
+                )
+        }
+    }
+
+    override fun checkBalance(paymentMethodData: PaymentMethodDetails) {
+        launch {
+            sessionRepository.checkBalance(paymentMethodData)
+                .fold(
+                    onSuccess = { response ->
+                        if (response.balance.value <= 0) {
+                            onFailure("Not enough balance")
+                        } else {
+                            val balanceResult = BalanceResult(response.balance, response.transactionLimit)
+                            sendBalanceResult(BalanceDropInServiceResult.Balance(balanceResult))
+                        }
+                    },
+                    onFailure = ::onFailure
```

Comment:
The common `onFailure` method will always return `DropInServiceResult.Error` but that only applies for payments and details calls. For example in this method we should be returning `BalanceDropInServiceResult.Error`. Internally it will work because we handle all errors the same way but if we change this handling later we might face unexpected issues.

Reply:
Ah right, I'll change it to the specific types